### PR TITLE
Add jemalloc compat functions.

### PR DESCRIPTION
**Pull request by davidchisnall**
This adds the full set of jemalloc functions that FreeBSD's libc
exposes, including some (the `*allocm` family) that are gone from newer
versions of jemalloc and the `*allocx` family that replaced them.  These
are not necessarily efficient implementations but they should allow
snmalloc to replace jemalloc without any ABI breakage (in the loosest
possible sense).

Jemalloc provides a very generic sysctl-like mechanism for setting and
getting some values.  These are all implemented to return the
not-supported error code.  This may break code that expects that they
will succeed.

In particular, these APIs are used to register custom backing-store
allocators and to manage caches and arenas.  These concepts don't map
directly onto snmalloc and attempting to do so would almost certainly
not provide the same performance characteristics and so it's better to
`LD_PRELOAD` jemalloc (or explicitly link to it) for programs that gain
a significant speedup from this.

### CODE REVIEW

**Comment by davidchisnall** [comment]
Note: We are currently missing tests for all of these.  I intend to fix that before it is merged.

File: src/test/func/jemalloc/jemalloc.cc
Comment (mjp41) [inline src/test/func/jemalloc/jemalloc.cc:108]: Could we use `SNMALLOC_CHECK`
https://github.com/microsoft/snmalloc/blob/main/src/ds/defines.h#L125

File: src/override/malloc.cc
Comment (mjp41) [inline src/override/malloc.cc:347]: Looking at the code size could still be below the size that is actually allocated by snmalloc.

snmalloc::aligned_size gives a size that when translated to a size class is guaranteed to meet the alignment, but it isn't necessarily the actual size.

```
snmalloc::aligned_size(16, 66) = 
   (66 -  1) | (16 - 1) + 1 = 80
```
But I think the next size class is 96. 
```suggestion
      *rsize = round_size(size);
```
should fix this.

File: src/override/malloc.cc
Comment (mjp41) [inline src/override/malloc.cc:365]: Not sure I understand the use of `extra` here?  Is the intended use, I need it to be `size` big, but if you need to reallocate, make it `extra` bigger.  I think your comment says that, but it took me a couple of passes.

File: src/override/malloc.cc
Comment (mjp41) [inline src/override/malloc.cc:411]:
```suggestion
        *rsize = alloc_size;
```

File: src/override/malloc.cc
Comment (mjp41) [inline src/override/malloc.cc:494]: Should we memset the difference?  Might be cheaper.  Though, not really important and keeping simple is good.

File: src/override/malloc.cc
Comment (mjp41) [inline src/override/malloc.cc:497]: Here you are using `size` which has been expanded by `aligned_size`, but not `round_size`.  This seems strange, you probably want the original size.

**Review by mjp41 — CHANGES_REQUESTED** [review]

**Comment by mjp41** [comment]
@davidchisnall I also wonder if this would be better in a 
```
override/jemalloc_compat.cc
```
or
```
override/freebsd_libc_extras.cc
```

Just to make it more separated out.  There is already quite a lot of functionality in `malloc.cc`.

File: src/override/malloc.cc
Comment (davidchisnall) [reply src/override/malloc.cc:365]: The actual requirement (from the half-sentence 'specification' in the jemalloc manual) for `extra` is that the caller would *like* `size + extra` bytes, but will accept anything above `size` bytes.  Since we're a sizeclass allocator, we always either return the current size if it's `>= size` or allocate to be at least `size+extra` here.

File: src/override/malloc.cc
Comment (davidchisnall) [reply src/override/malloc.cc:489]: This is copied from realloc.  I'm not actually sure what it is for.

File: src/test/func/jemalloc/jemalloc.cc
Comment (davidchisnall) [reply src/test/func/jemalloc/jemalloc.cc:265]: `malloc_usable_size` is spelled `sallocm` in this version of the API and is tested below.  This test runs with both our implementation and the one from jemalloc if it is also linked, so needs to have the malloc-implementation-specific things as template parameters.

File: src/test/func/jemalloc/jemalloc.cc
Comment (mjp41) [reply src/test/func/jemalloc/jemalloc.cc:108]: Can you improve SNMALLOC_CHECK?  It is only used in tests. I just would like to not have a proliferation of different bits for doing basically the same thing.

File: src/override/malloc.cc
Comment (mjp41) [inline src/override/malloc.cc:411]: Should this be:
```suggestion
        *rsize = round_size(alloc_size);
```

File: src/test/func/jemalloc/jemalloc.cc
Comment (mjp41) [reply src/test/func/jemalloc/jemalloc.cc:265]: I'm surprised the missing `round_size` didn't get caught by this check then?

File: src/override/malloc.cc
Comment (mjp41) [reply src/override/malloc.cc:493]: Agreed.  Let's leave.  Perhaps leave a comment for future readers, that there is a choice.

File: src/override/malloc.cc
Comment (mjp41) [reply src/override/malloc.cc:488]: Okay, I'll take a look at that code when I get some time.

File: src/test/func/jemalloc/jemalloc.cc
Comment (davidchisnall) [reply src/test/func/jemalloc/jemalloc.cc:268]: Not sure.  Possibly the sizes I test are always sufficiently close to a size class that the two are always the same?

File: src/test/func/jemalloc/jemalloc.cc
Comment (davidchisnall) [reply src/test/func/jemalloc/jemalloc.cc:111]: I don't think so easily.  `SNMALLOC_CHECK` calls `Pal::error`, which carefully avoids depending on `malloc`.  The `EXPECT` macro doesn't because it is never used in a situation where we expect `malloc` to be too broken for `fprintf` to work.  If we want to use `SNMALLOC_CHECK` for tests where the allocator under test is the process' `malloc` implementation then we can't make it do this.

Once we can depend on a C++ standard library that has `std::format` and `std::source_location` then we can make much nicer versions of both of these and so I'd prefer to limit the scope of a macro that uses `printf` format strings until we're willing to depend on a C++20 stdlib for the tests (even on platforms where we have a C++20 compiler, we don't always have a C++20 stdlib at the moment).  I could write a C++20 version that used `constexpr` if things to avoid memory allocation if the message didn't use use any format arguments, which would give the right behaviour in both cases, but it would break CI and a bunch of our not-tested-in-CI platforms.

I'm happy to do that in another PR that we don't merge until we're happy that platforms we care about have a sufficiently modern stdlib, if you like.

Comment (mjp41) [reply src/test/func/jemalloc/jemalloc.cc:111]: Okay. Makes sense.  Lets perhaps look at lifting something into `tests/utils.h` or something of that nature, so we can share across tests. (As part of a future PR.)

**Review by mjp41 — APPROVED** [review]

**Pull request merged**
